blobs.go

```diff
@@ -0,0 +1,64 @@
+package distribution
+
+import (
+	"io"
+	"time"
+
+	"github.com/docker/distribution/digest"
+)
+
+// Blob describes a readable and seekable binary object. Most registry
+// objects, including layers and manifests can be accessible by digest in a
+// BlobService.
+type Blob interface {
+	// NOTE(stevvooe): http.ServeContent requires an efficient implementation
+	// of ReadSeeker.Seek(0, os.SEEK_END).
+	io.ReadSeeker
+	io.ReaderAt
+	io.WriterTo
+	io.Closer
+
+	// Descriptor identifies the layer.
+	Descriptor() Descriptor
+}
+
+// BlobWriter provides a handle for working with a blob store. Instances
+// should be obtained from BlobService.Writer and BlobService.Resume. If
+// supported by the store, a writer can be recovered with the uuid.
+type BlobWriter interface {
+	io.WriteSeeker
+	io.WriterAt
+	io.ReaderFrom
+	io.Closer
+
+	// UUID returns the identifier for this writer.
+	UUID() string
+
+	// StartedAt returns the time this blob write was started.
+	StartedAt() time.Time
+
+	// Commit completes the blob writer process. The content is verified
+	// against the descriptor, which may result in an error. The only required
+	// field for the descriptor is the digest and mediatype. If the size is
+	// present, it will be checked as well.
+	Commit(desc Descriptor) (Blob, error)
```

Comment:
It would be convenient if all of Descriptor fields were optional and if omitted or empty the BlobWriter should use "application/octet-stream" as the default media type and just set the digest and size that it computed.